packages/did-resolver/src/base-resolver.ts

```diff
@@ -0,0 +1,92 @@
+import * as crypto from '@atproto/crypto'
+import { check } from '@atproto/common-web'
+import { AtprotoData, DidDocument, didDocument } from './types'
+import * as atprotoData from './atproto-data'
+import { DidNotFoundError, PoorlyFormattedDidDocumentError } from './errors'
+import { DidCache } from './did-cache'
+
+export abstract class BaseResolver {
+  constructor(public cache?: DidCache) {}
+
+  abstract resolveDidNoCheck(did: string): Promise<unknown | null>
+
+  validateDidDoc(did: string, val: unknown): DidDocument {
+    if (!check.is(val, didDocument)) {
+      throw new PoorlyFormattedDidDocumentError(did, val)
+    }
+    if (val.id !== did) {
+      throw new PoorlyFormattedDidDocumentError(did, val)
+    }
+    return val
+  }
+
+  async resolveDidNoCache(did: string): Promise<DidDocument | null> {
+    const got = await this.resolveDidNoCheck(did)
+    if (got === null) return null
+    return this.validateDidDoc(did, got)
+  }
+
+  async refreshCache(did: string): Promise<void> {
+    await this.cache?.refreshCache(did, () => this.resolveDidNoCache(did))
+  }
+
+  async resolveDid(
+    did: string,
+    forceRefresh = false,
+  ): Promise<DidDocument | null> {
+    if (this.cache && !forceRefresh) {
+      const fromCache = await this.cache.checkCache(did)
+      if (fromCache?.expired) {
+        this.refreshCache(did)
+      }
```

Comment:
I think that it can be useful to have separate ideas of expiration (not servable) versus staleness (servable but requires revalidation).  As things currently stand this item could be long-expired, and we'd still serve the current value.  I don't think we necessarily need to get into now and make any changes, but my inclination might be to have expiration working first for correctness, then layer-in staleness as an optimization.

Reply:
where do you draw the line between staleness & expired? staleness being "any cache hit" & expired being "older than TTL"?

Reply:
I usually think of it as two timeouts.  For example, in http it might look like this:
`Cache-Control: max-age=20, stale-while-revalidate=40`
 - for 0-20 seconds serve the cached version
 - for 20-60 seconds serve the cached version, and revalidate in the background
 - for 60+ seconds do not serve the cached version, revalidate and serve the latest result

Reply:
ahh got it got it 👍 

i'll take a pass 👍